test/unit/node/heart.test.ts

```diff
@@ -0,0 +1,98 @@
+import { logger } from "@coder/logger"
+import { readFile, writeFile } from "fs/promises"
+import { Heart, heartbeatTimer } from "../../../src/node/heart"
+import { clean, mockLogger, tmpdir } from "../../utils/helpers"
+
+function mockIsActive(resolveTo: boolean): () => Promise<boolean> {
+  return () => new Promise((resolve, reject) => setTimeout(() => resolve(resolveTo), 100))
+}
+
+describe("Heart", () => {
+  const testName = "heartTests"
+  let testDir = ""
+  let heart: Heart
+
+  beforeAll(async () => {
+    mockLogger()
+    await clean(testName)
+    testDir = await tmpdir(testName)
+  })
+  beforeEach(() => {
+    heart = new Heart(`${testDir}/shutdown.txt`, mockIsActive(true))
+  })
+  afterAll(() => {
+    jest.restoreAllMocks()
+  })
+  afterEach(() => {
+    jest.resetAllMocks()
+    if (heart) {
+      heart.dispose()
+    }
+  })
+  it("should write to a file when given a valid file path", async () => {
+    // Set up heartbeat file with contents
+    const text = "test"
+    const pathToFile = `${testDir}/file.txt`
+    await writeFile(pathToFile, text)
+    const fileContents = await readFile(pathToFile, { encoding: "utf8" })
+    expect(fileContents).toBe(text)
+
+    heart = new Heart(pathToFile, mockIsActive(true))
+    heart.beat()
+    // Check that the heart wrote to the heartbeatFilePath and overwrote our text
+    const fileContentsAfterBeat = await readFile(pathToFile, { encoding: "utf8" })
+    expect(fileContentsAfterBeat).not.toBe(text)
```

Comment:
One thing we might consider testing is that the timestamp updated (this is how to tell if code-server is active) rather than the contents (or in addition to the contents but the contents can technically be anything so idk if they need to be tested).

Reply:
Oh as in write a test that makes a call to the `/healthz` endpoint? I went with the simple approach for this one but agree, that would be worth testing later as well!

Reply:
Ah nah I mean something like (untested):

```typescript
const before = Date.now()
...
// Make sure the modified timestamp was updated.
const newStat = await fs.stat(pathToFile)
expect(newStat.modified).toBeGreaterThan(before)
```

Or something like that

Reply:
OH! I was wondering how you could do something like that. Let me play around with that.

Reply:
It worked! Had to modify slightly but you got me 90% of the way there 😎

```typescript
import { stat } from "fs/promises"
const before = Date.now()
heart.beat()
// Make sure the modified timestamp was updated.
const status = await fs.stat(pathToFile)
expect(status.mtimeMs).toBeGreaterThan(before)
```

Reply:
Awesome!  :tada:

Reply:
It runs so fast we might need greater than or equal lol

Reply:
I haven't had coffee yet so a little slow but realized that as well. Super odd? Getting before and after seems like the best solution to me. Thanks for calling that out!

Reply:
It could be that async issue, we might be issuing the stat before the write completes

Reply:
Oh right! Well...should I add that HACK I used into this test as well? I did run it locally and it passed, but it could be flakey.

Reply:
That could work, it is weird though because the content check passes so it does seem like the write finished.  :thinking:

Reply:
Astute observation! I almost forgot about that. Okay well if CI passes, I'm going to merge. If it fails, I'll add the hack.